File: lib/src/main/java/me/ilich/nestableviewpager/NestablePagerAdapterHelper.java
package me.ilich.nestableviewpager;

import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentPagerAdapter;
import android.support.v4.app.FragmentStatePagerAdapter;
import android.support.v4.view.PagerAdapter;
import android.support.v4.view.ViewPager;
import android.support.v7.app.AppCompatActivity;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;

public class NestablePagerAdapterHelper {

    public static void onCreateOptionsMenu(Menu menu, ViewPager viewPager, MenuInflater menuInflater) {
        menu.clear();
        fillIdList(viewPager, menu, menuInflater);
    }

    private static void fillIdList(ViewPager viewPager, Menu menu, MenuInflater menuInflater) {
        int currentPageIndex = viewPager.getCurrentItem();
        PagerAdapter adapter = viewPager.getAdapter();
        final Fragment fragment;
        if (adapter instanceof FragmentPagerAdapter) {
            fragment = ((FragmentPagerAdapter) adapter).getItem(currentPageIndex);
        } else if (adapter instanceof FragmentStatePagerAdapter) {
            fragment = ((FragmentStatePagerAdapter) adapter).getItem(currentPageIndex);
        } else {
            fragment = null;
        }
        if (fragment != null) {
            fragment.onCreateOptionsMenu(menu, menuInflater);
            if (fragment instanceof NestablePagerItem) {
                ViewPager nestedViewPager = ((NestablePagerItem) fragment).getNestedViewPager();
                if (nestedViewPager != null) {
                    fillIdList(nestedViewPager, menu, menuInflater);
                }
            }
        }
    }

    public static boolean onOptionsItemSelected(MenuItem menuItem, ViewPager viewPager) {
        boolean b = false;
        int currentPageIndex = viewPager.getCurrentItem();
        PagerAdapter adapter = viewPager.getAdapter();
        final Fragment fragment;
        if (adapter instanceof FragmentPagerAdapter) {
            fragment = ((FragmentPagerAdapter) adapter).getItem(currentPageIndex);
        } else if (adapter instanceof FragmentStatePagerAdapter) {
            fragment = ((FragmentStatePagerAdapter) adapter).getItem(currentPageIndex);
        } else {
            fragment = null;
        }
        if (fragment != null) {
            b = fragment.onOptionsItemSelected(menuItem);
            if (!b && fragment instanceof NestablePagerItem) {
                ViewPager nestedViewPager = ((NestablePagerItem) fragment).getNestedViewPager();
                if (nestedViewPager != null) {
                    b = onOptionsItemSelected(menuItem, nestedViewPager);
                }
            }
        }
        return b;
    }

    public static void addListenerToPager(ViewPager pager, final AppCompatActivity activity) {
        pager.addOnPageChangeListener(new ViewPager.OnPageChangeListener() {

            @Override
            public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {
                activity.supportInvalidateOptionsMenu();
            }

            @Override
            public void onPageSelected(int position) {

            }

            @Override
            public void onPageScrollStateChanged(int state) {

            }
        });
    }


}

File: lib/src/main/kotlin/me/ilich/nestableviewpager/NestablePagerAdapterHelper.kt
package me.ilich.nestableviewpager

import android.support.v4.app.FragmentPagerAdapter
import android.support.v4.view.ViewPager
import android.support.v7.app.AppCompatActivity
import android.view.Menu
import android.view.MenuInflater
import me.ilich.nestableviewpager.NestablePagerItem
import java.util.*

object NestablePagerAdapterHelper {

    fun onCreateOptionsMenu(menu: Menu, viewPager: ViewPager, menuInflater: MenuInflater) {
        menu.clear()
        val menuItemIds = ArrayList<Int>()
        fillIdList(viewPager, menuItemIds)
        for (menuItemId in menuItemIds) {
            menuInflater.inflate(menuItemId, menu)
        }
    }

    private fun fillIdList(viewPager: ViewPager, menuIdList: MutableList<Int>) {
        val currentPageIndex = viewPager.currentItem
        val adapter = viewPager.adapter as FragmentPagerAdapter
        val f = adapter.getItem(currentPageIndex)
        if (f is NestablePagerItem) {
            val items = f.getOptionsMenuIds()
            for (item in items) {
                menuIdList.add(item)
            }
            val nestedViewPager = f.getNestedViewPager()
            if (nestedViewPager != null) {
                fillIdList(nestedViewPager, menuIdList)
            }
        }
    }

    fun addListenerToPager(pager: ViewPager, activity: AppCompatActivity) {
        pager.addOnPageChangeListener(object : ViewPager.OnPageChangeListener {
            override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {
                activity.supportInvalidateOptionsMenu()
            }

            override fun onPageSelected(position: Int) {

            }

            override fun onPageScrollStateChanged(state: Int) {

            }
        })
    }

}
